File: frontend/src/lib/api/comments.ts
import { api } from "./client";

export interface CommentAuthor {
  id: string;
  username: string;
  display_name: string | null;
}

export interface Comment {
  id: string;
  parent_id: string | null;
  position_seconds: number | null;
  body_html: string;
  body_md: string;
  author: CommentAuthor;
  mentions: CommentAuthor[];
  edited_at: string | null;
  deleted_at: string | null;
  created_at: string;
}

export interface CommentsList {
  items: Comment[];
  total: number;
}

export const listComments = (slug: string) =>
  api<CommentsList>(`/api/sets/${encodeURIComponent(slug)}/comments`);

export const postComment = (
  slug: string,
  body: { body: string; parent_id?: string | null; position_seconds?: number | null },
) =>
  api<Comment>(`/api/sets/${encodeURIComponent(slug)}/comments`, {
    method: "POST",
    body: JSON.stringify(body),
  });

export const editComment = (id: string, body: string) =>
  api<Comment>(`/api/comments/${id}`, {
    method: "PATCH",
    body: JSON.stringify({ body }),
  });

export const deleteComment = (id: string) =>
  api<void>(`/api/comments/${id}`, { method: "DELETE" });

File: packages/core/src/setvault_core/schemas/comments.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CommentAuthor(BaseModel):
    id: str
    username: str
    display_name: str | None = None


class CommentOut(BaseModel):
    id: str
    parent_id: str | None = None
    position_seconds: int | None = None
    body_html: str
    body_md: str
    author: CommentAuthor
    mentions: list[CommentAuthor] = Field(default_factory=list)
    edited_at: datetime | None = None
    deleted_at: datetime | None = None
    created_at: datetime


class CommentCreateIn(BaseModel):
    body: str = Field(min_length=1, max_length=5000)
    parent_id: str | None = None
    position_seconds: int | None = Field(default=None, ge=0)


class CommentEditIn(BaseModel):
    body: str = Field(min_length=1, max_length=5000)


class CommentsListOut(BaseModel):
    items: list[CommentOut]
    total: int
